File: interview/longestWords.js
// Returns longest word in string
// See https://github.com/blakeembrey/code-problems/tree/master/problems/longest-words

function longest_words(str) {
  var max = 0,
      ret = [],
      len = str.length,
      cache = '',
      c = ''
        ;

  for (var i = 0; i < len; i++) {
    c = str[i];

    if (i == (len-1)) {
      cache += c;
      c = ' ';
    }
    if (c === ' ' ) {
      var l = cache.length;
      if (l > max) {
        ret = [];
        ret.push(cache);
        max = l;
      } else if (l === max){
        ret.push(cache);
      }
      cache = '';
    } else {
      cache += c;
    }
  }
  console.log(ret);
  return ret;
}

function longest_words(str) {
  var a = str.split(' '),
      max = 0,
      biggest = 0,
      ret = {};


  a.forEach(function(e) {
    var len = e.length;
    if ( max < len ) {
      max  = len;
      ret = {};
      ret[e.toLowerCase()] = len;
    } else if (max === len) {
      ret [e.toLowerCase()] = len;
    }
  });

  console.log (ret);

}


longest_words("You are just an old antidisestablishmentarian") // ["antidisestablishmentarian"]
longest_words("I gave a present to my parents") // ["present", "parents"]
longest_words("Buffalo buffalo Buffalo buffalo buffalo buffalo Buffalo buffalo") // ["buffalo"] or ["Buffalo"]

File: interview/LongestWords.java
// Write a function that returns the longest word(s) from a sentence. The
// function should not return any duplicate words (case-insensitive).
// https://github.com/blakeembrey/code-problems/tree/master/problems/longest-words
import java.util.*;

class LongestWords {
  public ArrayList<String> invoke(String phrase) {
    String[] words = phrase.split(" ");

    int max = 0;
    ArrayList<String> results = new ArrayList<String>();
    for (String word: words) {
      if (word.length() > max) {
        results = new ArrayList<String>();
        results.add(word.toLowerCase());
      }
      else if (word.length() == max && results.indexOf(word.toLowerCase()) > -1) {
        results.add(word.toLowerCase());
      }
    }
    return results;
  }

  public static void printArray(ArrayList<String> al) {
    System.out.println(al.toString());
  }
  public static void main(String[] args) {
    LongestWords lw = new LongestWords();
    printArray(lw.invoke(""));
    printArray(lw.invoke("Hi"));
    printArray(lw.invoke("Hi man"));
    printArray(lw.invoke("Hi man mAn"));
    printArray(lw.invoke("Hi man mAn hi hello"));
    printArray(lw.invoke("Hello Hi man mAn hi hello"));
  }
}

